replication_state.go

```diff
@@ -0,0 +1,320 @@
+	enabled  bool
+	logger   Logger
+	myNodeID NodeID
+
+	// receivedFinalizations maps sequences to a block and its associated finalization
+	seqs map[uint64]*finalizedQuorumRound
+
+	// rounds maps round numbers to QuorumRounds
+	rounds map[uint64]*QuorumRound
+
+	// digestTimeouts handles timeouts for fetching missing block digests.
+	// When a notarization depends on a block we haven’t received yet,
+	// it means a prior notarization for that block exists but is missing.
+	// Since we may not know which round that dependency belongs to,
+	// digestTimeouts ensures we re-request the missing digest until it arrives.
+	digestTimeouts *TimeoutHandler[Digest]
+
+	// emptyRoundTimeouts handles timeouts for fetching missing empty round notarizations.
+	// When replication encounters a notarized block that depends on an empty round we haven't received,
+	// emptyRoundTimeouts ensures we re-request those empty rounds until they are received.
+	emptyRoundTimeouts *TimeoutHandler[uint64]
+
+	// seqsToDigests stores a mapping of sequences to digests. It is used to remove old Digests from the digest timeout, when we have
+	seqsToDigests map[uint64]Digest
+
+	roundRequestor        *requestor
+	finalizationRequestor *requestor
+}
+
+func NewReplicationState(logger Logger, comm Communication, myNodeID NodeID, maxRoundWindow uint64, enabled bool, start time.Time, lock *sync.Mutex) *ReplicationState {
+	if !enabled {
+		return &ReplicationState{
+			enabled: enabled,
+			logger:  logger,
+		}
+	}
+
+	r := &ReplicationState{
+		enabled:  enabled,
+		myNodeID: myNodeID,
+		logger:   logger,
+
+		// seq replication
+		seqs:                  make(map[uint64]*finalizedQuorumRound),
+		seqsToDigests:         make(map[uint64]Digest),
+		finalizationRequestor: newRequestor(logger, start, lock, maxRoundWindow, comm, true),
+
+		// round replication
+		rounds:         make(map[uint64]*QuorumRound),
+		roundRequestor: newRequestor(logger, start, lock, maxRoundWindow, comm, false),
+	}
+
+	r.digestTimeouts = NewTimeoutHandler(logger, "digest", start, DefaultReplicationRequestTimeout, r.requestDigests, alwaysFalseRemover[Digest])
+	r.emptyRoundTimeouts = NewTimeoutHandler(logger, "empty", start, DefaultReplicationRequestTimeout, r.requestEmptyRounds, shouldRemoveFunc[uint64](shouldDelete))
+
+	return r
+}
+
+func (r *ReplicationState) AdvanceTime(now time.Time) {
+	if !r.enabled {
+		return
+	}
+
+	r.finalizationRequestor.advanceTime(now)
+	r.roundRequestor.advanceTime(now)
+	r.digestTimeouts.Tick(now)
+	r.emptyRoundTimeouts.Tick(now)
+}
+
+// deleteOldRounds cleans up the replication state for round replication after receiving a finalized round.
+func (r *ReplicationState) deleteOldRounds(finalizedRound uint64) {
+	for round := range r.rounds {
+		if round <= finalizedRound {
+			r.logger.Debug("Replication State Deleing Old Round", zap.Uint64("round", round))
+			delete(r.rounds, round)
+		}
+	}
+
+	r.roundRequestor.removeOldTasks(finalizedRound)
```

Comment:
what about `r.finalizationRequestor.removeOldTasks(finalizedRound)` ?  Why don't we need to do that too?